aten/src/THCUNN/generic/TemporalUpSamplingNearest.cu

```diff
@@ -5,151 +5,89 @@
+  }
+
+  if (gradOutput != NULL) {
+    THCUNN_check_dim_size(state, gradOutput, 3, 0, nBatch);
+    THCUNN_check_dim_size(state, gradOutput, 3, 1, nChannels);
+    THCUNN_check_dim_size(state, gradOutput, 3, 2, outputWidth);
   }
 }
 
 void THNN_(TemporalUpSamplingNearest_updateOutput)(
            THCState *state,
            THCTensor *input,
            THCTensor *output,
-           int scale_factor)
+           int outputWidth)
 {
-  THCTensor_(zero)(state, output);
-
   THCUNN_assertSameGPU(state, 2, input, output);
-  THNN_(TemporalUpSamplingNearest_shapeCheck)(state, input, NULL, scale_factor);
-  int inputWidth  = THCTensor_(size)(state, input,  input->dim()-1);
-  int outputWidth = inputWidth * scale_factor;
-
-   if (input->dim() == 2) {
-     THCTensor_(resize2d)(state, output,
-                          THCTensor_(size)(state, input, 0),
-                          outputWidth);
-   } else {
-     THCTensor_(resize3d)(state, output,
-                          THCTensor_(size)(state, input, 0),
-                          THCTensor_(size)(state, input, 1),
-                          outputWidth);
-  }
-
-  input = THCTensor_(newContiguous)(state, input);
-  // This is for allocating output Tensor
-  int64_t no_elements = 1;
-  for(int i = 0; i < input->dim(); i++){
-    no_elements *= input->size[i];
-  }
-  no_elements *= scale_factor;
+  int nbatch = THCTensor_(size)(state, input, 0);
+  int channels = THCTensor_(size)(state, input, 1);
+  int inputWidth  = THCTensor_(size)(state, input, 2);
 
-  int d1;
-  int d2;
+  THNN_(TemporalUpSamplingNearest_shapeCheck)(state, input, NULL, nbatch, channels, inputWidth, outputWidth);
+  THAssert(inputWidth > 0 && outputWidth > 0);
 
-  if (input->dim() == 2) {
-    d1 = output->size[0];
-    d2 = output->size[1];
-  } else {
-    d1 = output->size[1];
-    d2 = output->size[2];
-  }
 
-  real *input_data = THCTensor_(data)(state, input);
-  real *output_data = THCTensor_(data)(state, output);
-
-  // cuda blocks & threads:
-  int64_t nthreads = 256;
-  // Max number of blocks: http://en.wikipedia.org/wiki/CUDA
-  // 65535 for SM 2.x, 2^32 -1 for >= 3.0
-  // TODO: When we move to SM 3.5 we should update this
-  int64_t n_xblocks = min(max((int)ceil((float)no_elements / nthreads), 1), 65535);
-  int64_t n_yblocks = (int64_t)ceil((float)no_elements / (float)(n_xblocks * nthreads));
-  if (n_yblocks > 65535) {
-    THError("Input size is too large!  aborting");
-  }
-  dim3 blocks(n_xblocks, n_yblocks);
-  dim3 threads(nthreads);
+  THCTensor_(resize3d)(state, output,
+                       THCTensor_(size)(state, input, 0),
+                       THCTensor_(size)(state, input, 1),
+                       outputWidth);
+  THCTensor_(zero)(state, output);
 
-  // kernel:
-  upscale<<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (input_data, output_data, no_elements, scale_factor, d1, d2);
+  input = THCTensor_(newContiguous)(state, input);
```

Comment:
same question as before: because we are using the `THCDeviceTensor`, and we index all the tensors using the `[]` operator, do we need to keep the restriction that the input is contiguous?